pkg/backend/sqlite/driver.go

```diff
@@ -281,82 +333,88 @@ func NewDriver(ctx context.Context, config *DriverConfig) (*Driver, error) {
 // table if the key_value table exists, all in a single database transaction.
 // changes are rolled back if an error occurs.
-func setup(ctx context.Context, db database.Interface) error {
+func setup(ctx context.Context, db database.Interface) (SchemaVersion, error) {
+	var currentSchemaVersion SchemaVersion
 	conn, err := db.Conn(ctx)
 	if err != nil {
-		return err
+		return currentSchemaVersion, err
 	}
 	defer conn.Close()
 
 	// Optimistically ask for the user_version without starting a transaction
-	var currentSchemaVersion SchemaVersion
-
-	row := conn.QueryRowContext(ctx, `PRAGMA user_version`)
-	if err := row.Scan(&currentSchemaVersion); err != nil {
-		return err
+	currentSchemaVersion, err = getUserVersion(ctx, conn)
+	if err != nil {
+		return currentSchemaVersion, err
 	}
 
 	if err := currentSchemaVersion.CompatibleWith(databaseSchemaVersion); err != nil {
-		return err
+		// TODO: add a migration path for this case
+		// For now we will remain on the old schema version
+		return currentSchemaVersion, nil
 	}
 	if currentSchemaVersion >= databaseSchemaVersion {
-		return nil
+		return currentSchemaVersion, nil
 	}
 
 	txn, err := conn.BeginTx(ctx, nil)
 	if err != nil {
-		return err
+		return currentSchemaVersion, err
 	}
 	defer txn.Rollback()
 
-	if err := migrate(ctx, txn); err != nil {
-		return fmt.Errorf("migration failed: %w", err)
-	}
-
-	return txn.Commit()
-}
-
-// migrate tries to migrate from a version of the database
-// to the target one.
-func migrate(ctx context.Context, txn *sql.Tx) error {
-	var currentSchemaVersion SchemaVersion
-
-	row := txn.QueryRowContext(ctx, `PRAGMA user_version`)
-	if err := row.Scan(&currentSchemaVersion); err != nil {
-		return err
-	}
-
-	if err := currentSchemaVersion.CompatibleWith(databaseSchemaVersion); err != nil {
-		return err
-	}
 	if currentSchemaVersion >= databaseSchemaVersion {
-		return nil
+		return currentSchemaVersion, nil
 	}
 
-	switch currentSchemaVersion {
-	case NewSchemaVersion(0, 0):
+	if currentSchemaVersion == NewSchemaVersion(0, 0) {
 		if err := applySchemaV0_1(ctx, txn); err != nil {
-			return err
+			return currentSchemaVersion, fmt.Errorf("failed to apply schema v0.1: %w", err)
+		}
+		if err := applySchemaV1_0(ctx, txn); err != nil {
+			return currentSchemaVersion, fmt.Errorf("failed to apply schema v0.2: %w", err)
+		}
+		if setUserVersion(ctx, txn, databaseSchemaVersion); err != nil {
+			return currentSchemaVersion, fmt.Errorf("failed to set user version: %w", err)
 		}
-	default:
-		return nil
 	}
 
-	setUserVersionSQL := fmt.Sprintf(`PRAGMA user_version = %d`, databaseSchemaVersion)
-	if _, err := txn.ExecContext(ctx, setUserVersionSQL); err != nil {
-		return err
+	return currentSchemaVersion, txn.Commit()
+}
+
+func getUserVersion(ctx context.Context, conn *sql.Conn) (SchemaVersion, error) {
+	query := "PRAGMA user_version"
+
+	var userVersion SchemaVersion
+	var err error
+	retryCount := 0
```

Comment:
This is executed already in a "retry" context (the context above, line 323) so I think we should not retry here.